adapters/repos/db/vector/ssdhelpers/productQuantization.go

```diff
@@ -0,0 +1,254 @@
+	"math"
+	"sync"
+	"sync/atomic"
+
+	"github.com/weaviate/weaviate/adapters/repos/db/vector/hnsw/distancer"
+)
+
+type Encoder byte
+
+const (
+	UseTileEncoder   Encoder = 0
+	UseKMeansEncoder Encoder = 1
+)
+
+type DistanceLookUpTable struct {
+	calculated  [][]atomic.Bool
+	distances   [][]float32
+	center      []float32
+	segmentSize int
+}
+
+func NewDistanceLookUpTable(segments int, centroids int, center []float32) *DistanceLookUpTable {
+	distances := make([][]float32, segments)
+	calculated := make([][]atomic.Bool, segments)
+	for m := 0; m < segments; m++ {
+		distances[m] = make([]float32, centroids)
+		calculated[m] = make([]atomic.Bool, centroids)
+	}
+
+	return &DistanceLookUpTable{
+		distances:   distances,
+		calculated:  calculated,
+		center:      center,
+		segmentSize: len(center) / segments,
+	}
+}
+
+func (lut *DistanceLookUpTable) LookUp(
+	encoded []byte,
+	encoders []PQEncoder,
+	distanceProvider distancer.Provider,
+) float32 {
+	var sum float32
+
+	k := 0
+	for i, c := range encoded {
+		if lut.calculated[i][c].Load() {
+			sum += lut.distances[i][c]
+			k += lut.segmentSize
+		} else {
+			centroid := encoders[i].Centroid(c)
+			dist := float32(0)
+			for _, centroidX := range centroid {
+				dist += distanceProvider.Step(lut.center[k], centroidX)
+				k++
+			}
+			lut.distances[i][c] = dist
+			lut.calculated[i][c].Store(true)
+			sum += dist
+		}
+	}
+	return distanceProvider.Wrap(sum)
+}
+
+type ProductQuantizer struct {
+	ks                  int //centroids
+	m                   int //segments
+	ds                  int //dimensions per segment
+	distance            distancer.Provider
+	dimensions          int
+	kms                 []PQEncoder
+	encoderType         Encoder
+	encoderDistribution EncoderDistribution
+}
+
+type PQData struct {
+	Ks                  uint16
+	M                   uint16
+	Dimensions          uint16
+	EncoderType         Encoder
+	EncoderDistribution byte
+	Encoders            []PQEncoder
+}
+
+type PQEncoder interface {
+	Encode(x []float32) byte
+	Centroid(b byte) []float32
+	Add(x []float32)
+	Fit(data [][]float32) error
+	ExposeDataForRestore() []byte
+}
+
+// ToDo: Add a settings struct. Already necessary!!
+func NewProductQuantizer(segments int, centroids int, distance distancer.Provider, dimensions int, encoderType Encoder, encoderDistribution EncoderDistribution) *ProductQuantizer {
+	if dimensions%segments != 0 {
+		panic("dimension must be a multiple of m")
+	}
+	pq := &ProductQuantizer{
+		ks:                  centroids,
+		m:                   segments,
+		ds:                  int(dimensions / segments),
+		distance:            distance,
+		dimensions:          dimensions,
+		encoderType:         encoderType,
+		encoderDistribution: encoderDistribution,
+	}
+	return pq
+}
+
+func NewProductQuantizerWithEncoders(segments int, centroids int, distance distancer.Provider, dimensions int, encoderType Encoder, encoders []PQEncoder) *ProductQuantizer {
+	if dimensions%segments != 0 {
```

Comment:
This operation can panic if `segments == 0`

Reply:
True, the constructor should also return an error with the pointer so we handle the error instead of panicking. I also have an explicit panic somewhere else. I will change this. Thanks.